crates/bonsaidb-local/src/storage/argon.rs

```diff
@@ -0,0 +1,267 @@
+        saved_hash: String,
+    ) -> Result<(), Error> {
+        let (result_sender, result_receiver) = oneshot::channel();
+        if self
+            .sender
+            .send(HashRequest {
+                id,
+                password,
+                verify_against: Some(saved_hash),
+                result_sender,
+            })
+            .is_ok()
+        {
+            match result_receiver.await?.map_err(Error::from) {
+                Ok(_) => Ok(()),
+                Err(err) => {
+                    eprintln!("Error validating password for user {}: {:?}", id, err);
+                    Err(Error::Core(bonsaidb_core::Error::InvalidCredentials))
+                }
+            }
+        } else {
+            Err(Error::InternalCommunication)
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+struct HashingThread {
+    receiver: flume::Receiver<HashRequest>,
+    params: ArgonParams,
+    builder_template: Arc<OnceCell<Result<ParamsBuilder, ArgonError>>>,
+}
+
+impl HashingThread {
+    fn initialize_builder_template<R: Rng + CryptoRng>(
+        &self,
+        rng: &mut R,
+    ) -> Result<ParamsBuilder, ArgonError> {
+        match &self.params {
+            ArgonParams::Params(builder) => Ok(builder.clone()),
+            ArgonParams::Timed(config) => {
+                let mut params_builder = ParamsBuilder::new();
+                let params = params_builder
+                    .m_cost(config.ram_per_hasher)?
+                    .p_cost(config.threads_per_hasher)?
+                    .data(&0_u64.to_be_bytes())?;
+                let mut min_cost = 1;
+                let mut total_spent_t = 0;
+                let mut total_duration = Duration::ZERO;
+                let salt = SaltString::generate(rng);
+                loop {
+                    let t_cost = if total_spent_t > 0 {
+                        let average_duration_per_t = total_duration / total_spent_t;
+                        u32::try_from(ceil_divide(
+                            config.minimum_duration.as_nanos(),
+                            average_duration_per_t.as_nanos(),
+                        ))
+                        .unwrap()
+                        .max(min_cost)
+                    } else {
+                        min_cost
+                    };
+                    params.t_cost(t_cost)?;
+                    let params = params.clone().params()?;
+                    let argon = Argon2::new(Algorithm::Argon2id, Version::V0x13, params);
+
+                    let start = Instant::now();
+                    // TODO we should try to use into_memory but the problem is
+                    // that there's no way to calculate the number of blocks
+                    // needed through the public interface without
+                    // reimplementing a lot of logic. Although counterpoint: by
+                    // having it reallocate each time, hashing has a lower
+                    // persistent memory footprint.
```

Comment:
https://github.com/RustCrypto/password-hashes/pull/258
I will ask for a new release to be cut.